### ISSUE TITLE
BEEFY broken on Rococo after GRANDPA revert

### ISSUE THREAD

**Issue by serban300** [paritytech/polkadot-sdk#2285]
![image](https://github.com/paritytech/polkadot-sdk/assets/18124062/33a80399-288d-4eb0-8a82-8dee936cd7d3)

**Comment by serban300** [comment]
This is caused by some issues with `rococo-rpc-node-1`

It emits a lot of logs that go like this

`🥩 ran out of peers to request justif #7829924 from` logs

And also the `beefy_on_demand_justification_peer_refused` is very high lately

![image](https://github.com/paritytech/polkadot-sdk/assets/18124062/155f1da5-7387-41f6-bae8-a0bc114a2c99)

So probably the peers refuse to answer

**Comment by serban300** [comment]
Investigating on Adrian's test node which has the same issue as `rococo-rpc-node-1`

node identity: `12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a`

On the node there are a lot of logs like:
```
2023-11-14T10:01:20+02:00	2023-11-14 08:01:20.116 DEBUG tokio-runtime-worker beefy::sync: 🥩 requesting justif #7829924 from peer PeerId("12D3KooWDMqQL2m3ExdVpBXBmxBVF1R2HY9baF15NPVFNtyurpaq")    
2023-11-14T10:01:20+02:00	2023-11-14 08:01:20.116 DEBUG tokio-runtime-worker beefy::sync: 🥩 for on demand justification #7829924, peer PeerId("12D3KooWG8MczUuMUjXm4S1Y7eZeJuWhNZEQ7aHvAAkK2VuYsSYG") error: Refused    
2023-11-14T10:01:20+02:00	2023-11-14 08:01:20.115 DEBUG tokio-runtime-worker beefy: 🥩 vote target - mandatory block: #7829924    
2023-11-14T10:01:20+02:00	2023-11-14 08:01:20.115 DEBUG tokio-runtime-worker beefy::sync: 🥩 requesting justif #7829924 from peer PeerId("12D3KooWG8MczUuMUjXm4S1Y7eZeJuWhNZEQ7aHvAAkK2VuYsSYG")    
2023-11-14T10:01:20+02:00	2023-11-14 08:01:20.115 DEBUG tokio-runtime-worker beefy::sync: 🥩 for on demand justification #7829924, peer PeerId("12D3KooWM1vBvemiq7ZnSPNKWUvQTUGTDi6iQgfCrWKEsRjCzLFP") error: Refused    
2023-11-14T10:01:20+02:00	2023-11-14 08:01:20.113 DEBUG tokio-runtime-worker beefy: 🥩 vote target - mandatory block: #7829924 
```

So indeed the peers refuse to answer (probably they banned it).

And on the other validators I see a lot of logs like:
```
2023-11-10T15:10:42+02:00	2023-11-10 13:10:42.841 DEBUG tokio-runtime-worker parachain::reputation-aggregator: Reduce reputation peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") rep=CostMinor("Unexpected Statement, missing knowledge for relay parent")
2023-11-10T15:10:42+02:00	2023-11-10 13:10:42.835 DEBUG tokio-runtime-worker parachain::reputation-aggregator: Reduce reputation peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") rep=CostMinor("Unexpected Statement, missing knowledge for relay parent")
2023-11-10T15:10:42+02:00	2023-11-10 13:10:42.834 DEBUG tokio-runtime-worker parachain::reputation-aggregator: Reduce reputation peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") rep=CostMinor("Unexpected Manifest, missing knowlege for relay parent")
2023-11-10T15:10:42+02:00	2023-11-10 13:10:42.834 DEBUG tokio-runtime-worker parachain::statement-distribution: Received incoming manifest candidate_hash=0x3a0034aabb83c9a17eab575f6b4768d2163d9eeda07ece7ef9e94480df0c5518 peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") traceID=77096291967645156233897627732274538706
2023-11-10T15:10:42+02:00	2023-11-10 13:10:42.832 DEBUG tokio-runtime-worker parachain::reputation-aggregator: Reduce reputation peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") rep=CostMinor("Unexpected Statement, missing knowledge for relay parent")
```

```
2023-11-10T15:04:47+02:00	2023-11-10 13:04:47.746 DEBUG tokio-runtime-worker parachain::network-bridge-tx: peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") rep=ReputationChange { value: -4045000, reason: "Aggregated reputation change" } action="ReportPeer"
2023-11-10T15:04:47+02:00	2023-11-10 13:04:47.517 DEBUG tokio-runtime-worker parachain::network-bridge-tx: peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") rep=ReputationChange { value: -3980000, reason: "Aggregated reputation change" } action="ReportPeer"
2023-11-10T15:04:47+02:00	2023-11-10 13:04:47.375 DEBUG tokio-runtime-worker parachain::network-bridge-tx: peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") rep=ReputationChange { value: -3855000, reason: "Aggregated reputation change" } action="ReportPeer"
```

```
2023-11-10T15:04:18+02:00	2023-11-10 13:04:18.224 DEBUG tokio-runtime-worker parachain::reputation-aggregator: Reduce reputation peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") rep=CostMinorRepeated("Peer sent identical messages")
2023-11-10T15:04:18+02:00	2023-11-10 13:04:18.218 DEBUG tokio-runtime-worker parachain::reputation-aggregator: Reduce reputation peer=PeerId("12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a") rep=CostMinorRepeated("Peer sent identical messages")
```

**Comment by serban300** [comment]
Found this log on one of the validators:

```
2023-11-09 20:51:12.739  WARN tokio-runtime-worker sync: 💔 Block 0xea13975b16a70fb9727a7c0238920bc0df7c0c0bea0723af74ba6e41cccba299 received from peer 12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a has been blacklisted    
```

This is the block: https://polkadot.js.org/apps/#/explorer/query/0xea13975b16a70fb9727a7c0238920bc0df7c0c0bea0723af74ba6e41cccba299

This is block 7829923. Our nodes get stuck at block 7829924.

Probably this is what happened to all the peers and they dropped to connection with our nodes.

**Comment by acatangiu** [comment]
> Found this log on one of the validators:
> 
> ```
> 2023-11-09 20:51:12.739  WARN tokio-runtime-worker sync: 💔 Block 0xea13975b16a70fb9727a7c0238920bc0df7c0c0bea0723af74ba6e41cccba299 received from peer 12D3KooWHXsxWZxiBkNd1sFKSkyKTCxAPz6uoBQyGgvXZLKBdd3a has been blacklisted    
> ```
> 
> This is the block: https://polkadot.js.org/apps/#/explorer/query/0xea13975b16a70fb9727a7c0238920bc0df7c0c0bea0723af74ba6e41cccba299
> 
> This is block 7829923. Our nodes get stuck at block 7829924.
> 
> Probably this is what happened to all the peers and they dropped to connection with our nodes.

@altonen @andresilva any pointers where to go from here?

**Comment by andresilva** [comment]
That block was indeed added to `badBlocks` in the chain spec of the validators (maybe also to RPC nodes I don't know the details of the deployment). This was done when fixing rococo, since we had to revert two blocks and didn't want validators to immediately sync to the tip after reverting.

**Comment by serban300** [comment]
Ok, thanks for the details ! So sounds like this is not caused by a bug. It's just a side effect of blacklisting that block.

Then we can stop the investigation and just try to find a way to get the other nodes to accept `rococo-rpc-node-1`  back in the gossip

**Comment by serban300** [comment]
Actually still looking into the BEEFY issue.

It's strange that otherwise the chain advances on our nodes.

**Comment by serban300** [comment]
Took the value of the `PersistedState` of the BEEFY gadget from `rococo-adrian-node-0` and now I can reproduce the issue locally.

So the problem is not caused by the fact that other validators banned the node. It's something related to the `PersistedState`.

**Comment by serban300** [comment]
Could also reproduce the issue when syncing Rococo from scratch.

The problem is that the BEEFY gadget can't get the BEEFY justification for #7829924 from any peer. Probably what happened is the following:
- when fixing beefy a couple of days ago, a hard fork was created somewhere a bit before block #7829924
- BEEFY had already voted on the old #7829924
- for the nodes that were online at that point, the BEEFY gadget just moved on since it works only with block numbers
- for the nodes that were offline at that point, when they went back online later, the BEEFY gadget got stuck requesting the justification for block #7829924. That's because no one has it for the block on the hard fork (the canonical chain now). Everybody has the justification for the block #7829924 on the old fork

For fixing the issue, we should restart the BEEFY voting from before the hard fork for all validators. So for each validator we should:
- delete the `PersistedState` for the BEEFY gadget
- delete all the BEEFY justifications for the blocks after the hard fork

**Comment by acatangiu** [comment]
CC @andresilva @PierreBesson

@serban300 is right about the possible fixes, I see multiple options of going about it:

1. resync all validators (not practical)
2. resync one archive validator/node (`sync=fast` ok, warp sync not ok), which will also get stuck with BEEFY consensus at #7829924, take a DB snapshot, then restart all other validators from that snapshot
3. run a script on all/majority validators that scrubs BEEFY PersistedState and BEEFY justifications from the DB, then restart node - basically only resyncs/revotes BEEFY
4. run a modified `polkadot` binary on all/majority validators that has custom BEEFY voter that simply ignores existing BEEFY DB and resyncs/revotes BEEFY - then once BEEFY is caught up, switch back to normal `polkadot` binary

I think 4. is the easiest solution deployment-wise, 2. is the easiest one without custom logic

**Comment by andresilva** [comment]
I thought about what could have gone wrong here, we did revert some blocks but didn't revert anything that was finalized (that would have required all peers to resync the chain). So I'm still not sure how BEEFY got into the current state unless it acted on something that wasn't finalized.

Regarding solutions, can't we also reset BEEFY genesis on-chain?

**Comment by acatangiu** [comment]
> Regarding solutions, can't we also reset BEEFY genesis on-chain?

yes, I forgot to list it, that should work too. I'm ok with resetting BEEFY genesis to something after the mess

**Comment by serban300** [comment]
You're right. From the logs it looks like no finalized blocks were reverted. 

We did some more research and we found out why on most of the nodes the BEEFY gadget managed to advance. It's because they were restarted at block #7829924, but their state wasn't persisted correctly (the state was saved without #7829924). And after the restart they just skipped #7829924 and continued from the next mandatory block. Credits to @acatangiu for finding this. Fix here: https://github.com/paritytech/polkadot-sdk/pull/2378

We still don't know exactly why #7829924 wasn't finalized by BEEFY. Still looking on the logs.
